### PR TITLE
Yafi/reusable components

### PR DESCRIPTION
**Pull request by YafiAhsanH**
### Background: 
- created custom date picker component
- modified the form component to include the date picker
- moved the created components to their respective folders (forms and buttons)
- created register page and added its route to the config

### Additional Packages/Dependencies:
- intl (used to change date format for the date picker)

### Changes:
- common/components/forms/custom_date_picker.dart  (NEW)
- common/components/forms/custom_form_input.dart  (NEW)
- common/components/forms/custom_text_field.dart
- common/components/forms/form_component.dart
- page/register.dart
- common/config/route_generator.dart
- pubspec.yml

### How to Implement:
- [x] How to use the form component:
  
  - OLD:
    image
  
  - NEW:
  image
  
  Create instances of CustomFormInput class for each text field, No need to manually create the controllers for each text field, 
  each controller will have a built-in controller inside the CustomFormInput Class.

- [x] CustomFormInput parameters: 
  - label
  - type ("string" / "password" / "date")
  - errorMessage (Optional)

### CODE REVIEW

File: lib/common/components/forms/custom_date_picker.dart [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+class _CustomDatePickerState extends State<CustomDatePicker> {
+  DateTime selectedDate = DateTime.now();
+
+  Future<void> _selectDate(BuildContext context) async {
+    final DateTime? picked = await showDatePicker(
+      context: context,
+      initialDate: selectedDate,
+      firstDate: DateTime(2015),
```
Comment (RafadanaM) [inline lib/common/components/forms/custom_date_picker.dart]: I think this should be earlier like 100+ years from now or/and set a default value which can be changed

File: lib/common/components/forms/custom_date_picker.dart [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      lastDate: DateTime(2101),
```
Comment (RafadanaM) [inline lib/common/components/forms/custom_date_picker.dart]: This one too, give a default value, maybe current Date as a default value and make it changeable

File: lib/common/components/forms/custom_text_field.dart [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  State<CustomTextField> createState() => _CustomTextFieldState();
+}
+
+class _CustomTextFieldState extends State<CustomTextField> {
+  bool _error = false;
+
+  void validateField() {
+    setState(() {
+      _error =
+          widget.input.controller.text == 'a' && widget.input.errorMessage != ''
```
Comment (RafadanaM) [inline lib/common/components/forms/custom_text_field.dart]: what's a?

File: lib/page/register.dart [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+          } else {
+            return RegisterForm();
+          }
+        }),
+      ],
+    );
+  }
+}
+
+class RegisterForm extends StatelessWidget {
```
Comment (RafadanaM) [inline lib/page/register.dart:61]: I feel that register form should be in a different file, what do you think? @suryanirvana

**Review by RafadanaM — COMMENTED** [review]
I think there are some changes needed to be made

File: lib/page/register.dart (hunk 4 above)
Comment (suryanirvana) [reply lib/page/register.dart:61]: I agree

Comment (suryanirvana) [reply lib/page/register.dart:61]: Maybe think a way how you could generalise the form creation?

File: lib/page/register.dart [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  RegisterForm({Key? key}) : super(key: key);
+
+  final CustomFormInput name =
+      CustomFormInput(label: 'First Name', type: "string");
+  final CustomFormInput email =
+      CustomFormInput(label: 'Email', type: "string", errorMessage: "error");
+  final CustomFormInput password = CustomFormInput(
+      label: 'Password', type: "password", errorMessage: "error");
+  final CustomFormInput date =
+      CustomFormInput(label: 'Birth Date', type: "date", errorMessage: "error");
```
Comment (suryanirvana) [inline lib/page/register.dart:71]: missing several components

**Review by suryanirvana — CHANGES_REQUESTED** [review]
I noticed that the import still from "flutter_boilerplate", are there any ways to change that to something else because it's no longer a boilerplate😅

**Review by suryanirvana — APPROVED** [review]
looks-good-to-me

File: lib/common/components/forms/custom_date_picker.dart (hunk 2 above)
Comment (YafiAhsanH) [reply lib/common/components/forms/custom_date_picker.dart]: done fixing it, now we can use other values than the default for the first and last date

File: lib/common/components/forms/custom_text_field.dart (hunk 3 above)
Comment (YafiAhsanH) [reply lib/common/components/forms/custom_text_field.dart]: that's for testing the error message, it was supposed to check the regex there

Comment (YafiAhsanH) [reply lib/common/components/forms/custom_text_field.dart]: fixed it, now it checks the regex there

File: lib/page/register.dart (hunk 4 above)
Comment (RafadanaM) [reply lib/page/register.dart:61]: It's already generalised, look at the CustomForm component

**Review by RafadanaM — APPROVED** [review]
Looks good!

**Pull request merged**